// @inrupt/solid-client-authn-browser/dist/login/oidc/__mocks__/ClientRegistrar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mockDefaultClientRegistrar = exports.PublicClientRegistrarMock = exports.ClientRegistrarMock = exports.PublicClientRegistrarResponse = exports.ClientRegistrarResponse = void 0;
const globals_1 = require("@jest/globals");
exports.ClientRegistrarResponse = {
    clientId: "abcde",
    clientSecret: "12345",
};
exports.PublicClientRegistrarResponse = {
    clientId: "abcde",
};
exports.ClientRegistrarMock = {
    getClient: globals_1.jest.fn((_options, _issuerConfig) => Promise.resolve(exports.ClientRegistrarResponse)),
};
exports.PublicClientRegistrarMock = {
    getClient: globals_1.jest.fn((_options, _issuerConfig) => Promise.resolve(exports.PublicClientRegistrarResponse)),
};
const mockDefaultClientRegistrar = () => {
    return {
        getClient: globals_1.jest.fn((_options, _issuerConfig) => Promise.resolve(exports.ClientRegistrarResponse)),
    };
};
exports.mockDefaultClientRegistrar = mockDefaultClientRegistrar;
