// @inrupt/solid-client-authn-browser/dist/login/oidc/__mocks__/IssuerConfigFetcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mockIssuerConfigFetcher = exports.mockDefaultIssuerConfigFetcher = exports.IssuerConfigFetcherMock = exports.IssuerConfigFetcherFetchConfigResponse = void 0;
const globals_1 = require("@jest/globals");
exports.IssuerConfigFetcherFetchConfigResponse = {
    issuer: "https://idp.com",
    authorizationEndpoint: "https://idp.com/auth",
    tokenEndpoint: "https://idp.com/token",
    jwksUri: "https://idp.com/jwks",
    subjectTypesSupported: [],
    claimsSupported: [],
    grantTypesSupported: ["refresh_token"],
    idTokenSigningAlgValuesSupported: ["RS256"],
};
exports.IssuerConfigFetcherMock = {
    fetchConfig: globals_1.jest.fn((_issuer) => Promise.resolve(exports.IssuerConfigFetcherFetchConfigResponse)),
};
const mockDefaultIssuerConfigFetcher = () => {
    return {
        fetchConfig: globals_1.jest.fn().mockResolvedValue(exports.IssuerConfigFetcherFetchConfigResponse),
    };
};
exports.mockDefaultIssuerConfigFetcher = mockDefaultIssuerConfigFetcher;
const mockIssuerConfigFetcher = (config) => {
    return {
        fetchConfig: globals_1.jest.fn().mockResolvedValue(config),
    };
};
exports.mockIssuerConfigFetcher = mockIssuerConfigFetcher;
