// @inrupt/solid-client-authn-core/dist/authenticatedFetch/dpopUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateDpopKeyPair = exports.createDpopHeader = void 0;
const jose_legacy_modules_1 = require("@inrupt/jose-legacy-modules");
const uuid_1 = require("uuid");
function removeHashUsernameAndPassword(audience) {
    const cleanedAudience = new URL(audience);
    cleanedAudience.hash = "";
    cleanedAudience.username = "";
    cleanedAudience.password = "";
    return cleanedAudience.toString();
}
async function createDpopHeader(audience, method, dpopKey) {
    return new jose_legacy_modules_1.SignJWT({
        htu: removeHashUsernameAndPassword(audience),
        htm: method.toUpperCase(),
        jti: uuid_1.v4(),
    })
        .setProtectedHeader({
        alg: "ES256",
        jwk: dpopKey.publicKey,
        typ: "dpop+jwt",
    })
        .setIssuedAt()
        .sign(dpopKey.privateKey, {});
}
exports.createDpopHeader = createDpopHeader;
async function generateDpopKeyPair() {
    const { privateKey, publicKey } = await jose_legacy_modules_1.generateKeyPair("ES256");
    const dpopKeyPair = {
        privateKey,
        publicKey: await jose_legacy_modules_1.fromKeyLike(publicKey),
    };
    dpopKeyPair.publicKey.alg = "ES256";
    return dpopKeyPair;
}
exports.generateDpopKeyPair = generateDpopKeyPair;
