// @inrupt/solid-client-authn-core/dist/login/oidc/IClientRegistrar.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
const IClientRegistrar_1 = require("./IClientRegistrar");
globals_1.describe("handleRegistration", () => {
    globals_1.it("should perform DCR if a client WebID is provided, but the target IdP does not support Solid-OIDC", async () => {
        const options = {
            clientId: "https://some.app/registration#app",
            sessionId: "some session",
            tokenType: "DPoP",
        };
        const clientRegistrar = {
            getClient: globals_1.jest.fn(),
        };
        await IClientRegistrar_1.handleRegistration(options, { solidOidcSupported: undefined }, globals_1.jest.fn(), clientRegistrar);
        globals_1.expect(clientRegistrar.getClient).toHaveBeenCalled();
    });
    globals_1.it("should perform DCR if no client ID is provided", async () => {
        const options = {
            sessionId: "some session",
            tokenType: "DPoP",
        };
        const clientRegistrar = {
            getClient: globals_1.jest.fn(),
        };
        await IClientRegistrar_1.handleRegistration(options, { solidOidcSupported: undefined }, globals_1.jest.fn(), clientRegistrar);
        globals_1.expect(clientRegistrar.getClient).toHaveBeenCalled();
    });
    globals_1.it("should store provided client WebID if one provided and the Identity Provider supports Solid-OIDC", async () => {
        const options = {
            sessionId: "some session",
            tokenType: "DPoP",
            clientId: "https://my.app/registration#app",
        };
        const clientRegistrar = {
            getClient: globals_1.jest.fn(),
        };
        const storageUtility = {
            setForUser: globals_1.jest.fn(),
        };
        await IClientRegistrar_1.handleRegistration(options, {
            solidOidcSupported: "https://solidproject.org/TR/solid-oidc",
        }, storageUtility, clientRegistrar);
        globals_1.expect(clientRegistrar.getClient).not.toHaveBeenCalled();
        globals_1.expect(storageUtility.setForUser).toHaveBeenCalled();
    });
    globals_1.it("should store provided client registration information when the client ID is not a WebID", async () => {
        const options = {
            sessionId: "some session",
            tokenType: "DPoP",
            clientId: "some statically registered client ID",
            clientName: "some statically registered client name",
            clientSecret: "some statically registered client secret",
        };
        const clientRegistrar = {
            getClient: globals_1.jest.fn(),
        };
        const storageUtility = {
            setForUser: globals_1.jest.fn(),
        };
        await IClientRegistrar_1.handleRegistration(options, {
            solidOidcSupported: undefined,
        }, storageUtility, clientRegistrar);
        globals_1.expect(clientRegistrar.getClient).not.toHaveBeenCalled();
        globals_1.expect(storageUtility.setForUser).toHaveBeenCalled();
    });
});
globals_1.describe("determineSigningAlg", () => {
    globals_1.it("returns the preferred algorithm of the supported list", () => {
        globals_1.expect(IClientRegistrar_1.determineSigningAlg(["ES256", "HS256", "RS256"], ["ES256", "RS256"])).toBe("ES256");
        globals_1.expect(IClientRegistrar_1.determineSigningAlg(["ES256", "HS256", "RS256"], ["RS256"])).toBe("RS256");
        globals_1.expect(IClientRegistrar_1.determineSigningAlg(["RS256"], ["RS256"])).toBe("RS256");
    });
    globals_1.it("returns null if there are no matches", () => {
        globals_1.expect(IClientRegistrar_1.determineSigningAlg(["RS256"], ["ES256"])).toBeNull();
        globals_1.expect(IClientRegistrar_1.determineSigningAlg(["RS256"], [])).toBeNull();
    });
});
